### ISSUE TITLE
Event System

### ISSUE THREAD

**Issue by devpanda0** [yamp-project/v-issues#7]
**What is the scope of the api ?**
- [ ] World
- [ ] Ped
- [ ] Vehicle
- [x] General

**Propose an api design**

------------------------
The idea is not to need an extra RPC function if the event system could do it.

I built it with RabbitMQ and I find it more relaxed this way

As soon as `return` is included in the `.on*` function, it knows that it should be an "RPC".

I have 2 examples here of what I mean: 

Examples in JS/TS
Example 1 (with answer): 
```ts
yamp.onClient('example::event', (player: yamp.Player, example: string) => {
 return exmaple
})
```

Example 2 (no answer): 
```ts
yamp.onClient('example::event', (player: yamp.Player, example: string) => {
 console.log(example)
})

**Comment by ByAlexius** [comment]
Hey, thanks for your suggestion.
We are going to have an internal discussion on how to best implement this.

**Comment by szkiddaj** [comment]
How would you distinguish events using early returns (so it wouldn't be an if statement xmas tree) from RPCs?

**Comment by devpanda0** [comment]
> How would you distinguish events using early returns (so it wouldn't be an if statement xmas tree) from RPCs?

What do you mean exactly?

**Comment by szkiddaj** [comment]
> What do you mean exactly?

```ts
api.onClient("someRPCHandler", (player: api.Player, vehicleModel: string) => {
  // how would the api know, if it supposed to return the value given here, or just stop the execution?
  if (!auth.isPlayerAuthorized(player)) {
    return;
  }

  // it's a bad coding practive, but in case the developer
  // returns in the same line with a function call to an infobox, for example (or even more critical)
  if (!vehicleModel || !api.vehicleModels.includes(vehicleModel)) {
    return infobox.send(player, "Invalid vehicle model", "error");
  }
  
  // obv success return
  return true;
});
```

And there are probably more cases I didn't think about, that could cause problems.

**Comment by drakeee** [comment]
> > What do you mean exactly?
> 
> ```ts
> api.onClient("someRPCHandler", (player: api.Player, vehicleModel: string) => {
>   // how would the api know, if it supposed to return the value given here, or just stop the execution?
>   if (!auth.isPlayerAuthorized(player)) {
>     return;
>   }
> 
>   // it's a bad coding practive, but in case the developer
>   // returns in the same line with a function call to an infobox, for example (or it returns something it really shouldn't, more sensitive stuff.)
>   if (!vehicleModel || !api.vehicleModels.includes(vehicleModel)) {
>     return infobox.send(player, "Invalid vehicle model", "error");
>   }
>   
>   // obv success return
>   return true;
> });
> ```
> 
> And there are probably more cases I didn't think about, that could cause problems.

I guess the return value will be `undefined` in that case where there is no return value

**Comment by devpanda0** [comment]
yes

**Comment by devpanda0** [comment]
simply check whether a value is returned

**Comment by szkiddaj** [comment]
ehh idunno, I still think there would be edge cases with this.. It's better to be more strict, especially if a client could reach these handlers.

**Comment by devpanda0** [comment]
if you program cleanly, you won't have any difficulties

**Comment by ByAlexius** [comment]
Could you be more specific about what edge cases you mean?

**Comment by yannbcf** [comment]
```ts
yamp.onClient("name", (player: yamp.Player) => {
	return true;
}, true);
```

Another possibility could be to set an optional third argument to true when it is an rpc

**Comment by szkiddaj** [comment]
> if you program cleanly, you won't have any difficulties

Right, but what if a less experienced dev doesn't know how what's the difference between the two, and do something stupid?

> Could you be more specific about what edge cases you mean?

For example, the developer expects a handler to work as an RPC, but a (malicious) client triggers it as an event. (ik it's a dum example, but atm I couldn't think of any better)


I rather would do something like `yamp.onRPC`, so the developer could just look at the function call, and see if it's an event or an RPC.

**Comment by Yiin** [comment]
If we do the RPC it will for sure have different api from Events, it's way too confusing otherwise. 

something like 

events: emit, on
rpc: call, register
